core/state/access_witness.go

```diff
@@ -112,7 +112,7 @@ func (aw *AccessWitness) TouchAndChargeValueTransfer(callerAddr, targetAddr []by
 
 // TouchAndChargeContractCreateInit charges access costs to initiate
 // a contract creation
-func (aw *AccessWitness) TouchAndChargeContractCreateInit(addr []byte, createSendsValue bool) uint64 {
+func (aw *AccessWitness) TouchAndChargeContractCreateInit(addr []byte) uint64 {
```

Comment:
I think this is unfortunatly a bit more complex than that: if the contract already exists (so there's a collision) AND we don't send value, then we should not write to it.

Reply:
Are you referring to the `, true)` in L117? If that's the case I agree.

Reply:
yes, that makes sense

Reply:
A write and a fill

Reply:
i have a other issues in mind now, since we always wirte the codeHash when creating a account, this `ContractCreateInit` and `ContractCreateCompleted` as seperate functions does not make sense to me.
since we always write to the codeHash (except for collision check) we should remove the `ContractCreateCompleted` event and do everything at the start only.

Reply:
yeah but we write a different type of hash if that is the case, and no code. And that write of the hash only happens if the account doesn't exist.

Reply:
yes, but we write something, and in ContractCreateInit we dont charge codeHash write costs, we only charge BasicDataLeaf, which seems incorrect to me because we are reading and writing to both the field.
again this argument is assuming that account does not exists at all.
if we find a collision then it makes sense to charge the cost of read/write after the contract creation is completed. But in case where no collison exists and the create call fails, so we dont end up calling ContractCreateCompleted, but we still write the basic data leaf and code hash leaf with a empty value.